# HG changeset 9744 patch
From: kaf24@firebug.cl.cam.ac.uk
# Node ID 02159733ec97aaeed6edaebecdc0e3b95d2e2dea
# Parent  6ce2965720ee2ed191700834102e867009ea14ce
Subject: Only destroy a machine-contiguous memory region if
it really is contiguous (e.g., create_contiguous did not fail).
Signed-off-by: Keir Fraser <keir@xensource.com>
xen-unstable changeset:   10353:bd1a0b2bb2d4596227951ad6d36cb4fcc2d00a8e
xen-unstable date:        Thu Jun 15 11:35:23 2006 +0100

Acked-by: Jan Beulich <jbeulich@novell.com>

Index: head-2006-06-13/arch/i386/mm/hypervisor.c
===================================================================
--- head-2006-06-13.orig/arch/i386/mm/hypervisor.c	2006-06-16 10:57:12.000000000 +0200
+++ head-2006-06-13/arch/i386/mm/hypervisor.c	2006-06-16 11:00:13.000000000 +0200
@@ -365,7 +365,8 @@ void xen_destroy_contiguous_region(unsig
 		.domid        = DOMID_SELF
 	};
 
-	if (xen_feature(XENFEAT_auto_translated_physmap))
+	if (xen_feature(XENFEAT_auto_translated_physmap) ||
+	    !test_bit(__pa(vstart) >> PAGE_SHIFT, contiguous_bitmap))
 		return;
 
 	scrub_pages(vstart, 1 << order);
